DisCatSharp.Hosting/DiscordSharedHostedService.cs

```diff
@@ -0,0 +1,84 @@
+// This file is part of the DisCatSharp project, a fork of DSharpPlus.
+//
+// Copyright (c) 2021 AITSYS
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+using System;
+using System.Threading.Tasks;
+using DisCatSharp.Configuration;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace DisCatSharp.Hosting
+{
+    /// <summary>
+    /// Simple Implementation for <see cref="DiscordShardedClient"/> to work as a <see cref="BackgroundService"/>
+    /// </summary>
+    public abstract class DiscordShardedHostedService : BaseHostedService, IDiscordHostedShardService
+    {
+        public DiscordShardedClient ShardedClient { get; protected set; }
+
+#pragma warning disable 8618
+        protected DiscordShardedHostedService(IConfiguration config,
+            ILogger<DiscordShardedHostedService> logger,
+            IServiceProvider serviceProvider,
+            IHostApplicationLifetime applicationLifetime,
+            string configBotSection = DisCatSharp.Configuration.ConfigurationExtensions.DefaultRootLib)
+            : base(config, logger, serviceProvider, applicationLifetime, configBotSection)
+        {
+
+        }
+#pragma warning restore 8618
+
+        protected override Task PreConnectAsync()
+        {
+            try
+            {
+                var config = this.Configuration.ExtractConfig<DiscordConfiguration>("Discord", this.BotSection);
```

Comment:
Why are important functions now in a method that will be overridden in most cases? Imo, this is not convenient, because it obliges you to always refer to this method using `base.PreConnectAsync();`. Maybe I'm wrong...

Reply:
> 
> 
> LGTM, but unfortunately, in this case, the use of services in events such as SocketOpened is blocked, since it is fired immediately after the client is initialized, so even before the services are copied into shards, this event will already be triggered. Idk if anything can be done in this case

What if we just pass the provider in through the configuration process? Like ApplicationCommands does? Then when the client is created we should be able to immediately pull the provider from the config instead of  creating client THEN manually setting the provider?

Reply:
> 
> 
> Why are important functions now in a method that will be overridden in most cases? Imo, this is not convenient, because it obliges you to always refer to this method using `base.PreConnectAsync();`. Maybe I'm wrong...

That's a valid point.